File: 3.The sum of three nums.py
'''
   @FilePath            : \AlgorithmSlay\3.The sum of three nums.py
   @Author              : AClearZhang
   @Date                : 2020-11-19 21:55:24
   @E-mail              : aclearzhang@qq.com
   @Homepage            : www.aclear.top
   @LastEditors         : AClearZhang
   @LastEditTime        : 2020-11-20 21:19:48
   @Version             : 1.0
   @Description         :  注意以下题干描述

    给你一个包含 n 个整数的数组 nums，判断 nums 中是否存在三个元素 a，b，c ，使得 a + b + c = 0 ？请你找出所有满足条件且不重复的三元组。
    注意：答案中不可以包含重复的三元组。
    示例：

    给定数组 nums = [-1, 0, 1, 2, -1, -4]，

    满足要求的三元组集合为：
    [
    [-1, 0, 1],
    [-1, -1, 2]
    ]

'''
from typing import List
# class Solution:
#     '''
#        @description: 
#        @param {*} self
#        @param {List} nums
#        @return {*}
#        @notes: 寻找三元组。返回下标
#     '''
#     def threeSum(self, nums: List[int]) -> List[List[int]]:
#         # 先排序 再双指针寻找
#         # sort_num = sorted(range(len(nums)),key=lambda k: nums[k])
#         sort_num = sorted(nums)
#         answer = []
#         # answer = ()
#         head = 0
#         # tail = len(nums) - 1
#         while head < len(nums) - 2:
#             tail = len(nums) - 1
#             if head!=0 and sort_num[head] == sort_num[head-1]: #多次添加
#                 head+=1
#                 while sort_num[head] == sort_num[head-1] and head!=tail-1:
#                     head+=1
#             while tail > head + 1:
               
#                 if (tail!=len(nums)-1) and sort_num[tail] == sort_num[(tail+1)%len(nums)]:
#                     tail -= 1
#                     while sort_num[tail] == sort_num[(tail+1)%len(nums)] and tail!=head+1:
#                         tail -= 1
#                 #print("now head:{},tail:{}, sort_num:{}, answer:{}".format(head,tail, sort_num, answer))
#                 other = 0 - (sort_num[head] + sort_num[tail])
#                 if other > 0:
#                     mid = tail - 1
#                     while mid != head:
#                         if sort_num[mid] < other:
#                             # 无了
#                             tail = tail - 1
#                             break
#                         elif sort_num[mid] > other:
#                             mid = mid - 1
#                         elif sort_num[mid] == other:
#                             # 对了
#                             answer.append([sort_num[head], sort_num[mid], sort_num[tail]])
#                            # print(answer)
#                             tail = tail - 1
#                             break
#                 elif other < 0:
#                     mid = head + 1
#                     while mid != tail:
#                         if sort_num[mid] < other:
#                             mid+=1
#                         elif sort_num[mid] > other:
#                             # 无了
#                             tail-=1
#                             break
#                         elif sort_num[mid] == other:
#                             # 对了
#                             answer.append([sort_num[head], sort_num[mid], sort_num[tail]])
#                             #print(answer)
#                             tail-=1
#                             break
#                 elif other == 0:
#                     mid = head + 1
#                     while mid != tail:
#                         if sort_num[mid] == 0:
#                             answer.append([sort_num[head], 0, sort_num[tail]])
#                             #print(answer)
#                             # tail-=1
#                             break
#                         mid+=1
#                     tail-=1
#             head+=1
#         return answer


# Solution2：
# 关键之处在于边界，以及去重
from typing import List
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        if( not nums or n < 3):
            return []
        res = []
        sort_num = sorted(nums)
        for i,v in enumerate(sort_num):
            if(sort_num[i] > 0): # 遗漏了！如果当前数字大于0，则三数之和一定大于0，所以结束循环
                return res
            if i >= 1 and sort_num[i] == sort_num[i-1]: #去重
                continue
            L = i+1; R = n-1
            while L<R:
                sum = sort_num[i] + sort_num[L] + sort_num[R]
                if sum == 0:
                    res.append([sort_num[i], sort_num[L], sort_num[R]])
                    # 去重并寻找下一个不是次数的数值
                    L+=1; R-=1; # 可以这样 因为排好序了，已经不可能只移动一边了！ YES！
                    while(L<R and sort_num[L] == sort_num[L-1]): L+=1
                    while(R>L and sort_num[R] == sort_num[R+1]): R-=1
                elif sum>0:
                    R-=1
                elif sum < 0 :
                    L+=1

                # # 去重L&R
                # while(L<R and sort_num[L+1] == sort_num[L]): L+=1
                # while(L<R and sort_num[R    -1] == sort_num[R]): R-=1
                # sum = sort_num[i] + sort_num[L] + sort_num[R]
                # if sum==0:
                #     res.append([sort_num[i],sort_num[L],sort_num[R]])
                # elif sum > 0:
                #     R-=1
                # elif sum < 0:
                #     L+=1
        return res
                





if __name__ == "__main__":
    a = Solution()
    # print(a.threeSum([-1,0,1,2,-1,-4]))
    # print(a.threeSum([0,0,0,0]))
    print(a.threeSum([1,0, -1]))
    # print(a.threeSum([-4,-2,-2,-2,0,1,2,2,2,3,3,4,4,6,6]))

File: C++/3.The sum of three nums.cpp
/*
 * @FilePath            : \AlgorithmSlay\C++\3.The sum of three nums.cpp
 * @Author              : AClearZhang
 * @Date                : 2020-11-19 21:59:12
 * @E-mail              : aclearzhang@qq.com
 * @Homepage            : www.aclear.top
 * @LastEditors         : AClearZhang
 * @LastEditTime        : 2020-11-23 22:37:28
 * @Version             : 1.0
 * @Description         : C++ 实现题目三

给你一个包含 n 个整数的数组 nums，判断 nums 中是否存在三个元素 a，b，c ，使得 a + b + c = 0 ？请你找出所有满足条件且不重复的三元组。

注意：答案中不可以包含重复的三元组。

 

示例：

给定数组 nums = [-1, 0, 1, 2, -1, -4]，

满足要求的三元组集合为：
[
  [-1, 0, 1],
  [-1, -1, 2]
]
 
 */
#include <iostream>
#include <vector>
#include <windows.h>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
      // 三数之和； 关键是边界限制，防止重复！
      // 首先写出算法思想
     
     
      // 首先 无数或者<3返回null
      // 然后 排序数组
      // head = 0
      // 最后：for all
      //         > 0 returen
      //         jump, 

      sort(nums.begin(), nums.end());
      if(nums.empty() || nums.size() < 3){
        return vector<int> [];
      }

    }
};


int main(){
  Solution s;
  vector<int> c = {2, 7, 11, 15};
  vector<int> a = s.twoSum(c, 9);
  for (int i = 0; i < a.size(); i++)
  {
      cout << a[i] << " ";
  }
  printf("Hello World!\n");
  system("pause");
  return 0;
}



// #include <stdio.h>
// #include <iostream>
// #include <vector>
// #include <set>
// #include <algorithm>

// using namespace std;


// /*
//  *   Similar like "Two Number" problem, we can have the simlar solution.
//  *
//  *   Suppose the input array is S[0..n-1], 3SUM can be solved in O(n^2) time on average by 
//  *   inserting each number S[i] into a hash table, and then for each index i and j,  
//  *   checking whether the hash table contains the integer - (s[i]+s[j])
//  *
//  *   Alternatively, the algorithm below first sorts the input array and then tests all 
//  *   possible pairs in a careful order that avoids the need to binary search for the pairs 
//  *   in the sorted list, achieving worst-case O(n^n)
//  *
//  *   Solution:  Quadratic algorithm
//  *   http://en.wikipedia.org/wiki/3SUM
//  *
//  */
// vector<vector<int> > threeSum(vector<int> &num) {

//     vector< vector<int> > result;
//     if(num.size() == 0 || num.size() == 1 || num.size() == 2) return result;

//     //sort the array, this is the key
//     sort(num.begin(), num.end());

//     int n = num.size();

//     for (int i=0; i<n-2; i++) {
//         //skip the duplication
//         if (i > 0 && num[i - 1] == num[i]) continue;
//         int a = num[i];
//         int low = i + 1;
//         int high = n - 1;
//         while (low < high) {
//             int b = num[low];
//             int c = num[high];
//             if (a + b + c == 0) {
//                 //got the soultion
//                 vector<int> v;
//                 v.push_back(a);
//                 v.push_back(b);
//                 v.push_back(c);
//                 result.push_back(v);
//                 // Continue search for all triplet combinations summing to zero.
//                 //skip the duplication
//                 while(low < n - 1 && num[low] == num[low + 1]) low++; 
//                 while(high > 0 && num[high] == num[high - 1]) high--; 
//                 low++;
//                 high--;
//             } else if (a+b+c > 0) {
//                 //skip the duplication
//                 while(high > 0 && num[high] == num[high - 1]) high--;
//                 high--;
//             } else {
//                 //skip the duplication
//                 while(low < n - 1 && num[low] == num[low + 1]) low++;
//                 low++;
//             } 
//         }
//     }
//     return result;
// }

// //using combination method could meet <<Time Limit Exceeded>> error
// vector<vector<int> > combination(vector<int> &v, int k);
// bool isSumZero(vector<int>& v);
// int sum(vector<int>& v);

// vector<vector<int> > threeSum2(vector<int> &num) {
//     vector< vector<int> > result;
//     vector< vector<int> > r = combination(num, 3);
//     for (int i = 0; i < r.size(); i++) {
//         if (isSumZero(r[i])) {
//             result.push_back(r[i]);
//         }
//     }
//     return result;
// }

// bool isSumZero(vector < int>& v) {
//     return sum(v) == 0;
// }

// int sum(vector<int>& v) {
//     int s = 0;
//     for(int i = 0; i < v.size(); i++) {
//         s += v[i];
//     }
//     return s;
// }

// vector<vector<int> > combination(vector<int> &v, int k) {

//     vector<vector<int> > result;
//     vector<int> d;
//     int n = v.size();
//     for (int i = 0; i < n; i++) {
//         d.push_back( (i < k) ? 1 : 0 );
//     }

//     //1) from the left, find the [1,0] pattern, change it to [0,1]
//     //2) move all of the 1 before the pattern to the most left side
//     //3) check all of 1 move to the right
//     while(1) {
//         vector<int> tmp;
//         for(int x = 0; x < n; x++) {
//             if (d[x]) tmp.push_back(v[x]);
//         }
//         sort(tmp.begin(), tmp.end());
//         result.push_back(tmp);
//         //step 1), find [1,0] pattern
//         int i;
//         bool found = false;
//         int ones = 0;
//         for(i = 0; i < n - 1; i++) {

//             if (d[i] == 1 && d[i + 1] == 0) {
//                 d[i] = 0; d[i + 1] = 1;
//                 found = true;
//                 //step 2) move all of right 1 to the most left side
//                 for (int j = 0; j < i; j++) {
//                     d[j] = ( ones > 0 ) ? 1 : 0;
//                     ones--;
//                 }
//                 break;
//             }
//             if (d[i] == 1) ones++;
//         }
//         if (!found) {
//             break;
//         }

//     }
//     return result;
// }


// void printMatrix(vector<vector<int> > &matrix)
// {
//     for(int i = 0; i < matrix.size(); i++) {
//         printf("{");
//         for(int j = 0; j < matrix[i].size(); j++) {
//             printf("%3d ", matrix[i][j]) ;
//         }
//         printf("}\n");
//     }
//     cout << endl;
// }


// int main()
// {
//     //int a[] = { -1, 0, 1, 2, -1, 1, -4 };
//     int a[] = { -1, 1, 1, 1, -1, -1, 0,0,0 };
//     vector<int> n(a, a + sizeof(a)/sizeof(int));
//     vector< vector<int> > result = threeSum(n);
//     printMatrix(result);    
//     return 0;
// }